CValidationInterface Cleanups

TheBlueMatt opened the pull request:
This + #9605 + about three more and wallet updates from blocks will be in a background thread so that they do not block block connection. This PR is just some general cleanup that happened along the way or otherwise makes future changes easier. See commit messages for more details.

TheBlueMatt committed:
Remove dead-code tracking of requests for blocks we generated

TheBlueMatt committed:
Use std::shared_ptr instead of boost::shared_ptr in ScriptForMining

TheBlueMatt committed:
Include missing #include in zmqnotificationinterface.h

TheBlueMatt committed:
Add override to functions using CValidationInterface methods

sdaftuar reviewed (approved):
One comment nit in the zmq file, utACK f3ef07215b

TheBlueMatt commented on src/zmq/zmqnotificationinterface.cpp:
```diff
@@ -144,8 +144,10 @@ void CZMQNotificationInterface::UpdatedBlockTip(const CBlockIndex *pindexNew, co
 }
 
-void CZMQNotificationInterface::SyncTransaction(const CTransaction& tx, const CBlockIndex* pindex, int posInBlock)
+void CZMQNotificationInterface::TransactionAddedToMempool(const CTransactionRef& ptx)
```
OK, added a few comments.

ryanofsky replied:
In commit "SyncTransaction->TxAddedToMempool/BlockConnected/Disconnected"

>> Perhaps we should have a comment here

> OK, added a few comments.

FWIW, not seeing a comment here. Maybe they were added elsewhere, but it would seem good to mention right in the top of the function body that this is not only called when a transaction is added to the mempool like the name would suggest. Alternately, you could leave this function named SyncTransaction and have TransactionAddedToMempool call SyncTransaction to avoid the misleading title.

TheBlueMatt replied:
Added another comment.

theuni commented on src/zmq/zmqnotificationinterface.cpp:
```diff
@@ -160,3 +162,19 @@ void CZMQNotificationInterface::SyncTransaction(const CTransaction& tx, const CB
     }
 }
+
+void CZMQNotificationInterface::BlockConnected(const std::shared_ptr<const CBlock>& pblock, const CBlockIndex* pindexConnected, const std::vector<CTransactionRef>& vtxConflicted)
```
To help others review, these two functions seem to mesh with current behavior. I don't see how they could possibly be usable by any client, but there ya go.

theuni commented on src/wallet/wallet.cpp:
```diff
@@ -1181,11 +1181,10 @@ void CWallet::MarkConflicted(const uint256& hashBlock, const uint256& hashTx)
+    const CTransaction& tx = *ptx;
 
-    if (!AddToWalletIfInvolvingMe(tx, pindex, posInBlock, true))
+    if (!AddToWalletIfInvolvingMe(tx, pindexBlockConnected, posInBlock, true))
```
Mind changing AddToWalletIfInvolvingMe to take a CTransactionRef here? Looks like we end up keeping 2 copies otherwise.

TheBlueMatt replied:
Fixed.

theuni commented on src/validation.cpp:
```diff
@@ -2518,14 +2511,10 @@ bool ActivateBestChain(CValidationState &state, const CChainParams& chainparams,
-                const CBlock& block = *(pair.second);
-                for (unsigned int i = 0; i < block.vtx.size(); i++)
-                    GetMainSignals().SyncTransaction(*block.vtx[i], pair.first, i);
+                GetMainSignals().BlockConnected(pair.second, pair.first, *mrt.GetConflictedTransactions());
```
The same set of conflicted transactions is sent here for each block.

TheBlueMatt replied:
Fixed, with a lot more code :(.

TheBlueMatt committed:
Add pblock to connectTrace at the end of ConnectTip, not start

This makes ConnectTip responsible for the ConnectTrace instead
of splitting the logic between ActivateBestChainStep and ConnectTip

TheBlueMatt committed:
Make ConnectTrace::blocksConnected private, hide behind accessors

TheBlueMatt committed:
Handle conflicted transactions directly in ConnectTrace

theuni commented on src/wallet/wallet.cpp:
```diff
@@ -1198,6 +1197,38 @@ void CWallet::SyncTransaction(const CTransaction& tx, const CBlockIndex *pindex,
+        SyncTransaction(ptx, NULL, -1);
+    }
+    for (size_t i = 0; i < pblock->vtx.size(); i++) {
+        SyncTransaction(pblock->vtx[i], pindex, pindex ? i : -1);
```
pindex can no longer be NULL here.

TheBlueMatt replied:
Fixed.

TheBlueMatt committed:
Keep conflictedTxs in ConnectTrace per-block

theuni commented on src/validation.cpp:
```diff
@@ -2518,14 +2511,10 @@ bool ActivateBestChain(CValidationState &state, const CChainParams& chainparams,
-
             // Transactions in the connected block are notified
             for (const auto& pair : connectTrace.blocksConnected) {
                 assert(pair.second);
```
Assert pair.first here too, I think. It'd be nice to doc that the BlockConnected() params are now always non-null.

TheBlueMatt replied:
Fixed.

theuni commented:
Concept ACK.

TheBlueMatt committed:
Take a CTransactionRef in AddToWalletIfInvolvingMe to avoid a copy

TheBlueMatt commented:
To fix the issue @theuni found ("The same set of conflicted transactions is sent here for each block."), I had to go rewrite a big chunk of how the mempool removal callback logic works. Now what was previously a standalone class which listens to mempool for transactions removed has gone into the ConnectTrace class, which is accessed in ActivateBestChain to generate callbacks.

TheBlueMatt commented:
Rebased for trivial conflict from #9605.

ryanofsky commented on src/net_processing.h:
```diff
@@ -30,10 +30,10 @@ class PeerLogicValidation : public CValidationInterface {
-    virtual void UpdatedBlockTip(const CBlockIndex *pindexNew, const CBlockIndex *pindexFork, bool fInitialDownload);
-    virtual void BlockChecked(const CBlock& block, const CValidationState& state);
-    virtual void NewPoWValidBlock(const CBlockIndex *pindex, const std::shared_ptr<const CBlock>& pblock);
+    virtual void BlockConnected(const std::shared_ptr<const CBlock>& pblock, const CBlockIndex* pindexConnected, const std::vector<CTransactionRef>& vtxConflicted) override;
```
In commit "Add override to functions using CValidationInterface methods"

I think normally you drop "virtual" when you add "override" to make the two different types of method declarations more distinct. At least that's the style I'm used to, and we seem to follow it everywhere else we're currently using override.

TheBlueMatt replied:
OK, done.

ryanofsky reviewed (commented):
utACK d9c8399716c6688689f02da41105588ae0d72c16.

Looks good, but suggested possible improvements.

TheBlueMatt commented on src/validation.cpp:
```diff
@@ -2195,6 +2195,12 @@ static int64_t nTimeFlush = 0;
 static int64_t nTimePostConnect = 0;
 
+struct PerBlockConnectTrace {
+    CBlockIndex* pindex = NULL;
```
Is there actually a difference between nullptr and NULL to the compiler (it overrides differently, sometimes, if you're calling a function, no?)

bulldozer00 replied:
From Cpp Core Guidelines (https://github.com/isocpp/CppCoreGuidelines/blob/master/CppCoreGuidelines.md#Res-nullptr)

ES.47: Use nullptr rather than 0 or NULL ()

Readability. Minimize surprises: nullptr cannot be confused with an int. nullptr also has a well-specified (very restrictive) type, and thus works in more scenarios where type deduction might do the wrong thing on NULL or 0.

TheBlueMatt commented on src/validation.cpp:
```diff
@@ -2195,6 +2195,12 @@ static int64_t nTimeFlush = 0;
+struct PerBlockConnectTrace {
+    CBlockIndex* pindex = NULL;
+    std::shared_ptr<const CBlock> pblock;
+    std::shared_ptr<std::vector<CTransactionRef> > conflictedTxs;
```
Heh, I still find the pre-C++11 version to be easier to read :P

TheBlueMatt commented on src/validation.cpp:
```diff
@@ -2240,24 +2240,23 @@ struct ConnectTrace {
- * The block is always added to connectTrace (either after loading from disk or by copying
- * pblock) - if that is not intended, care must be taken to remove the last entry in
- * blocksConnected in case of failure.
+ * The block is always added to connectTrace in the case connection succeeds.
```
Tweaked, carryover from too-little-diff-change.

TheBlueMatt commented on src/validation.cpp:
```diff
@@ -2233,7 +2233,17 @@ static int64_t nTimePostConnect = 0;
     std::vector<std::pair<CBlockIndex*, std::shared_ptr<const CBlock> > > blocksConnected;
+
+public:
+    void BlockConnected(CBlockIndex* pindex, const std::shared_ptr<const CBlock>& pblock) {
```
I prefer to force peoplt to not take the atomic increment hit without changing the code :).

TheBlueMatt replied:
Ahh, indeed, I hate this C++11 stuff, anyway, I took the proposed change.

TheBlueMatt commented on src/validation.cpp:
```diff
@@ -2233,7 +2233,17 @@ static int64_t nTimePostConnect = 0;
+        blocksConnected.emplace_back(pindex, pblock);
+    }
+
+    std::vector<std::pair<CBlockIndex*, std::shared_ptr<const CBlock> > > GetBlocksConnected() {
```
Oops, indeed. Now it just returns a reference, which should also be fine. The method is not technically const later (without making things mutable), so I'll leave the second part.

TheBlueMatt commented on src/validation.cpp:
```diff
@@ -2291,6 +2290,8 @@ bool static ConnectTip(CValidationState& state, const CChainParams& chainparams,
     LogPrint("bench", "  - Connect postprocess: %.2fms [%.2fs]\n", (nTime6 - nTime5) * 0.001, nTimePostConnect * 0.000001);
     LogPrint("bench", "- Connect block: %.2fms [%.2fs]\n", (nTime6 - nTime1) * 0.001, nTimeTotal * 0.000001);
+
+    connectTrace.blocksConnected.emplace_back(pindexNew, pthisBlock);
```
Sure, added more to the commit message.

TheBlueMatt committed:
Handle SyncTransaction in ActivateBestChain instead of ConnectTrace

This makes a later change to move it all into one per-block callback
simpler.

TheBlueMatt commented on src/validation.cpp:
```diff
@@ -2200,17 +2200,18 @@ static int64_t nTimePostConnect = 0;
 
 public:
     ConnectTrace(CTxMemPool &_pool) : pool(_pool) {
+        conflictedTxs.emplace_back();
```
OK, did that for the whole patch set.

TheBlueMatt commented on src/validation.cpp:
```diff
@@ -2220,28 +2226,33 @@ class ConnectTrace {
-        blocksConnected.emplace_back(pindex, pblock);
-        conflictedTxs.emplace_back();
+        assert(!awaitingClear);
+        assert(pindex && pblock);
```
Done.

TheBlueMatt commented on src/validation.cpp:
```diff
@@ -2220,28 +2226,33 @@ class ConnectTrace {
+
+    std::vector<PerBlockConnectTrace> GetBlocksConnected() {
+        if (!blocksConnected.back().pindex) {
+            blocksConnected.pop_back();
```
Done.

TheBlueMatt commented on src/validation.cpp:
```diff
@@ -2220,28 +2226,33 @@ class ConnectTrace {
+        if (!blocksConnected.back().pindex) {
+            blocksConnected.pop_back();
+            assert(!awaitingClear);
+            awaitingClear = true;
```
Sorry, the intention of making conflictedTxs a shared_ptr is so that it can be easily passed to CValidationInterface in a later patchset, not so that we can avoid copy here, forgot to mention that when I split up the patch set.

TheBlueMatt commented on src/validation.cpp:
```diff
@@ -2231,19 +2198,46 @@ static int64_t nTimePostConnect = 0;
+
+    void NotifyEntryRemoved(CTransactionRef txRemoved, MemPoolRemovalReason reason) {
+        if (reason == MemPoolRemovalReason::CONFLICT) {
+            conflictedTxs.push_back(txRemoved);
```
Indeed, done.

TheBlueMatt commented on src/wallet/wallet.cpp:
```diff
@@ -1172,6 +1171,38 @@ void CWallet::SyncTransaction(const CTransaction& tx, const CBlockIndex *pindex,
+
+void CWallet::BlockConnected(const std::shared_ptr<const CBlock>& pblock, const CBlockIndex *pindex, const std::vector<CTransactionRef>& vtxConflicted) {
+    LOCK2(cs_main, cs_wallet);
+    // TODO: Tempoarily ensure that mempool removals are notified before
```
Its removed later in the PR, so probably easier to just leave it

TheBlueMatt commented on src/validation.cpp:
```diff
@@ -2526,8 +2512,15 @@ bool ActivateBestChain(CValidationState &state, const CChainParams& chainparams,
             // throw all transactions though the signal-interface
-
-            } // MemPoolConflictRemovalTracker destroyed and conflict evictions are notified
+            connectTrace.CallSyncTransactionOnConflictedTransactions();
```
It all still happens under cs_main? There are no behavior changes there. I added another line to the commit message.

TheBlueMatt replied:
Argh, I believe that could be a race which triggers a reachable assertion if there are two ActivateBestChain's going at once, moved the class into the cs_main lock, which simplified the whole thing a bit.

TheBlueMatt commented on src/wallet/wallet.cpp:
```diff
@@ -1155,11 +1155,10 @@ void CWallet::MarkConflicted(const uint256& hashBlock, const uint256& hashTx)
 
-void CWallet::SyncTransaction(const CTransaction& tx, const CBlockIndex *pindex, int posInBlock)
-{
-    LOCK2(cs_main, cs_wallet);
```
There are no behavior changes here? Everything is still under cs_main from the previous commit.

ryanofsky replied:
> > The wallet now no longer releases cs_main and cs_wallet between transactions when it is processing multiple transactions from the same block.

> There are no behavior changes here? Everything is still under cs_main from the previous commit.

Everything is still under cs_main, but previously cs_main and cs_wallet were released between processing individual transactions in the block, while now they are held until all transactions in the block are processed. You already mention this in your commit message, I was just suggesting moving the locking change to a separate commit so this one can be a pure refactoring, and the locking part can be understood on its own (and more easily identified and reverted if it causes problems).

TheBlueMatt replied:
Ahh, OK, split the commit.

ryanofsky replied:
In commit "SyncTransaction->TxAddedToMempool/BlockConnected/Disconnected":

This commit is currently removing LOCK2 at the top of SyncTransaction and then copying it 4 times over each place where SyncTransaction is called. Not a big deal, I guess, since the locks move again in the next commit. But just to be clear, what I was trying to suggest was only moving the locks in the _next_ commit, and leaving them completely alone in this commit.

TheBlueMatt replied:
Ahh, oops, whatever, I'll just leave it.

TheBlueMatt commented on src/validationinterface.h:
```diff
@@ -33,7 +34,9 @@ void UnregisterAllValidationInterfaces();
 class CValidationInterface {
```
CValidationInterface has never listened to mempool, it was just an interface between the things in validation.cpp and other callers, now its also listening to mempool, putting it in the middle of the various callbacks we have going.

ryanofsky replied:
> CValidationInterface has never listened to mempool, it was just an interface between the things in validation.cpp and other callers, now its also listening to mempool, putting it in the middle of the various callbacks we have going.

I guess I don't understand what "listening to mempool" means. CValidationInterface is a class with methods, and the methods seem to be called from the same places as before. I don't think it's an important point, I just don't get what this is trying to say.

TheBlueMatt replied:
Ahh, I see the issue. Yes, this is actually referring to something further in the change-set, but not in this PR (the full changeset is at https://github.com/TheBlueMatt/bitcoin/commits/2017-01-wallet-cache-inmempool, this is referring to https://github.com/TheBlueMatt/bitcoin/commit/71683dc0d8a00e16211194aaa66bfea1ae265f45).

TheBlueMatt commented:
Added a bunch of comments and additional commit description at @ryanofsky's request. Also, note that the reason the std::vector<CTransactionRef> of conflicted transactions is passed back to ActivateBestChain as a shared_ptr is to make it easy to add that to a queue of events shared between threads later, not for anything in this function, forgot to make that clear when I mentioned this is part of a much longer patchset.

ryanofsky commented on src/validation.cpp:
```diff
@@ -2219,28 +2232,40 @@ class ConnectTrace {
+        // awaitingClear to keep track of this state, and pop the last
+        // entry here to make sure the list we return is sane.
+        if (!blocksConnected.back().pindex) {
+            blocksConnected.pop_back();
```
In commit "Handle SyncTransaction in ActivateBestChain instead of ConnectTrace"

Could / should you `assert(blocksConnected.back().conflictedTxs.empty())` here?

TheBlueMatt replied:
Took this change (as a part of larger changes).

ryanofsky commented on src/validation.cpp:
```diff
@@ -2219,28 +2232,40 @@ class ConnectTrace {
+        // entry here to make sure the list we return is sane.
+        if (!blocksConnected.back().pindex) {
+            blocksConnected.pop_back();
+            assert(!awaitingClear);
```
In commit "Handle SyncTransaction in ActivateBestChain instead of ConnectTrace"

What do you think of my suggestion to get merge the ClearBlocksConnected and GetBlocksConnected methods into a single method so you can get rid of the awaitingClear variable and all the asserts? 

Link: https://github.com/bitcoin/bitcoin/pull/9725#discussion_r104711671

Seems better to prevent the interface misuse instead of detecting it at runtime and crashing.

TheBlueMatt replied:
Made the class use-once instead.

ryanofsky reviewed (commented):
utACK 80e7290f5b7c030c5bb2f2f108783d64732bbd10

Thanks for making changes, especially like the new comments.

TheBlueMatt committed:
Hold cs_wallet for whole block [dis]connection processing

This simplifies fixing the wallet-returns-stale-info issue as we
now hold cs_wallet across an entire block instead of only per-tx.

ryanofsky reviewed (commented):
utACK 96370a99bc3d3728df11a4dc289514dfa6ccfcfc.

Thanks for the updates.

TheBlueMatt commented:
Rebased to fix trivial conflict in rpc/mining.cpp

JeremyRubin commented on src/validation.cpp:
```diff
@@ -2175,17 +2175,17 @@ static int64_t nTimePostConnect = 0;
 private:
     std::vector<std::pair<CBlockIndex*, std::shared_ptr<const CBlock> > > blocksConnected;
-    std::vector<CTransactionRef> conflictedTxs;
+    std::vector<std::vector<CTransactionRef> > conflictedTxs;
```
nit: extra whitespace

TheBlueMatt replied:
Done

JeremyRubin commented on src/validation.cpp:
```diff
@@ -2203,15 +2204,18 @@ class ConnectTrace {
+                GetMainSignals().SyncTransaction(*tx, NULL, CMainSignals::SYNC_TRANSACTION_NOT_IN_BLOCK);
+            }
         }
         conflictedTxs.clear();
```
Worth doing a shrink_to_fit/swap on empty vector here?

```c++

std::vector<std::vector<CTransactionRef>>(1).swap(conflictedTxs);

```

TheBlueMatt replied:
Hmm, its removed by the end and there is no equivalent vector on which to do so, so I think its fine.

JeremyRubin commented on src/validation.cpp:
```diff
@@ -2194,29 +2207,32 @@ class ConnectTrace {
+        assert(pindex);
+        assert(pblock);
+        blocksConnected.back().pindex = pindex;
+        blocksConnected.back().pblock = std::move(pblock);
```
I think given where we call BlockConnected you could make pblock be an rvalue and move from the caller, saving the pass by copy.

TheBlueMatt replied:
Likely, but the copy-from-rvalue is mostly free, and as you previously pointed out the caller now does std::move, so I dont think it matters.

JeremyRubin commented on src/validation.cpp:
```diff
@@ -2194,29 +2207,32 @@ class ConnectTrace {
+        assert(!blocksConnected.back().pindex);
+        assert(blocksConnected.back().conflictedTxs->empty());
+        blocksConnected.pop_back();
         return blocksConnected;
```
Can you set a flag so that doing things with the instance after calling GetBlocksConnected is an assert(blocks_not_gotten)?

Otherwise, could you return a pair of iterators instead?

TheBlueMatt replied:
Its already there implicitly, and I was trying to pair back the million asserts I had in this code (because if you have NotifyEntryRemoved(), you will assert the pindex on the back is empty, and BlockConnected will assert that its setting a pindex, so if you try to call either it will fail as the back().pindex is no longer null).

JeremyRubin replied:
Maybe worth explicitly differentiating the cause though -- back() being null could be caused for two reasons, and if you call twice with 0 blocksconnected it will cause UB on the second call. Also, that @sdaftuar found a double call so it's worth having this.

TheBlueMatt replied:
Wait, where does it UB? I think it always asserts if you misuse it. Anyway, no need to make asserts overly complicated, the preconditions here are well-documented, the asserts are probably just as useful to remove completely as leave there.

JeremyRubin commented on src/rpc/mining.cpp:
```diff
@@ -207,7 +206,7 @@ UniValue generatetoaddress(const JSONRPCRequest& request)
         throw JSONRPCError(RPC_INVALID_ADDRESS_OR_KEY, "Error: Invalid address");
 
-    boost::shared_ptr<CReserveScript> coinbaseScript(new CReserveScript());
+    std::shared_ptr<CReserveScript> coinbaseScript(new CReserveScript());
```
please use make_shared

TheBlueMatt replied:
Done.

JeremyRubin commented on src/wallet/wallet.cpp:
```diff
@@ -3296,9 +3296,9 @@ void CWallet::UpdatedTransaction(const uint256 &hashTx)
+void CWallet::GetScriptForMining(std::shared_ptr<CReserveScript> &script)
 {
-    boost::shared_ptr<CReserveKey> rKey(new CReserveKey(this));
+    std::shared_ptr<CReserveKey> rKey(new CReserveKey(this));
```
make_shared

TheBlueMatt replied:
Done

JeremyRubin commented on src/validationinterface.h:
```diff
@@ -42,7 +41,7 @@ class CValidationInterface {
     virtual void ResendWalletTransactions(int64_t nBestBlockTime, CConnman* connman) {}
     virtual void BlockChecked(const CBlock&, const CValidationState&) {}
-    virtual void GetScriptForMining(boost::shared_ptr<CReserveScript>&) {};
+    virtual void GetScriptForMining(std::shared_ptr<CReserveScript>&) {};
```
nit: could refactor API to pass up an rvalue reference rather than take an arg.

TheBlueMatt replied:
I'll leave that for someone else in a later PR.

JeremyRubin commented on src/wallet/wallet.cpp:
```diff
@@ -958,8 +958,9 @@ bool CWallet::LoadToWallet(const CWalletTx& wtxIn)
  * posInBlock signals or by checking mempool presence when necessary.
  */
-bool CWallet::AddToWalletIfInvolvingMe(const CTransaction& tx, const CBlockIndex* pIndex, int posInBlock, bool fUpdate)
+bool CWallet::AddToWalletIfInvolvingMe(const CTransactionRef& ptx, const CBlockIndex* pIndex, int posInBlock, bool fUpdate)
```
You probably don't need to use a ref to a CTransactionRef because CWalletTx is going to make a copy anyways -- you could just move it from the CWalletTx constructor.

TheBlueMatt replied:
CWalletTx now copies the ref that is passed in, so using a ref here should avoid a complete copy?

JeremyRubin replied:
The CWalletTx constructor is:
```c++
CWalletTx(const CWallet* pwalletIn, CTransactionRef arg) : CMerkleTx(std::move(arg))
```

I was suggesting to add a constructor

```c++
CWalletTx(const CWallet* pwalletIn, CTransactionRef&& arg) : CMerkleTx(arg)
```

and then update L984 to

```c++
            CWalletTx wtx(this, std::move(ptx));
```

and update this line to

```c++
bool CWallet::AddToWalletIfInvolvingMe(const CTransactionRef ptx, const CBlockIndex* pIndex, int posInBlock, bool fUpdate)
```

There are two reasons:
  - slightly more clear intent (AddToWalletIfInvolvingMe does get ownership)
  - Eliminates an indirection

TheBlueMatt replied:
Ahh, OK, I missed your original point. Anyway, I would still prefer to leave it because it is a smaller diff to not change CWalletTx, and the way it is you dont end up with a variable lying around which has already been std::move()d prior to the end of the function, which I dont super like doing. Anyway, feel free to clean it up in a future PR (also maybe we need a general "when to && vs const& vs copy-and-move style guide)

JeremyRubin reviewed (commented):
I've reviewed most of this and it seems reasonable. I left some feedback on some minor ways to improve it, but I don't think anything is a hold up.

TheBlueMatt commented on src/zmq/zmqnotificationinterface.h:
```diff
@@ -8,6 +8,7 @@
 #include "validationinterface.h"
 #include <string>
 #include <map>
+#include <list>
```
Yes, there is an std::list in CZMQNotificationInterface. It might not fail build due to current header organization, but it should be included if its used (and I believe it failed build when I was making changes).

JeremyRubin replied:
Ah, I was viewing the diff and didn't realize it was truncated, so I didn't see the list :) ACK the include.

TheBlueMatt commented on src/validation.cpp:
```diff
@@ -2215,24 +2215,23 @@ struct ConnectTrace {
- * blocksConnected in case of failure.
+ * The block is added to connectTrace if connection succeeds.
  */
 bool static ConnectTip(CValidationState& state, const CChainParams& chainparams, CBlockIndex* pindexNew, const std::shared_ptr<const CBlock>& pblock, ConnectTrace& connectTrace)
```
Well the other thing is I'd like to move the new mempool disconnect cache stuff into ConnectTrace, at which point ConnectTip again will end up needing to know about the connecttrace object. Anyway, agreed, lets table this for future work.

TheBlueMatt commented on src/validation.cpp:
```diff
@@ -2266,6 +2265,8 @@ bool static ConnectTip(CValidationState& state, const CChainParams& chainparams,
     LogPrint("bench", "  - Connect postprocess: %.2fms [%.2fs]\n", (nTime6 - nTime5) * 0.001, nTimePostConnect * 0.000001);
     LogPrint("bench", "- Connect block: %.2fms [%.2fs]\n", (nTime6 - nTime1) * 0.001, nTimeTotal * 0.000001);
+
+    connectTrace.blocksConnected.emplace_back(pindexNew, pthisBlock);
```
Done.

sipa commented:
utACK acac363e574470be7ca26ab43b7c32c2beac9d9c.

sdaftuar commented on src/validation.cpp:
```diff
@@ -2206,19 +2173,46 @@ static int64_t nTimePostConnect = 0;
  * part of a single ActivateBestChainStep call.
+ *
+ * This class also tracks transactions that are removed from the mempool as
+ * conflicts and and can be used to pass all those transactions through
```
nit: "and and"

TheBlueMatt replied:
Fixed.

sdaftuar commented on src/validation.cpp:
```diff
@@ -2502,12 +2515,20 @@ bool ActivateBestChain(CValidationState &state, const CChainParams& chainparams,
-            for (const auto& pair : connectTrace.GetBlocksConnected()) {
-                assert(pair.second);
-                const CBlock& block = *(pair.second);
+            for (const PerBlockConnectTrace& trace : connectTrace.GetBlocksConnected()) {
```
This is a second call to GetBlocksConnected(), which violates the single-use requirement.

TheBlueMatt replied:
OK, I'll just squash a few commits before merge, I dont particularly feel strongly about there being so many commits and dont want to have to reorder commits to fix this.

TheBlueMatt replied:
Fixed.

sdaftuar reviewed (commented):
Looks like one of the intermediate commits is broken (but the issue goes away in a later commit).

TheBlueMatt committed:
SyncTransaction->TxAddedToMempool/BlockConnected/Disconnected

This simplifies fixing the wallet-returns-stale-info issue as we
can now hold cs_wallet across an entire block instead of only
per-tx (though we only actually do so in the next commit).

This change also removes the NOT_IN_BLOCK constant in favor of only
passing the CBlockIndex* parameter to SyncTransactions when a new
block is being connected, instead of also when a block is being
disconnected.

This change adds a parameter to BlockConnectedDisconnected which
lists the transactions which were removed from mempool due to
confliction as a result of this operation. While its somewhat of a
shame to make block-validation-logic generate a list of mempool
changes to be included in its generated callbacks, fixing this isnt
too hard.

Further in this change-set, CValidationInterface starts listening
to mempool directly, placing it in the middle and giving it a bit
of logic to know how to route notifications from block-validation,
mempool, etc (though not listening for conflicted-removals yet).

kallewoof commented:
utACK 312683b23752bd9bdae365d0695f1ed73d3c1f3b

sipa commented:
Needs rebäse.

TheBlueMatt commented:
Rebased trivially.

JeremyRubin commented:
utack 461e49f

theuni commented:
cautious utACK b1a6d4cd560fbdb66506841860db03c08ea4bbbc.

jnewbery commented on src/wallet/wallet.cpp:
```diff
@@ -966,8 +966,9 @@ bool CWallet::LoadToWallet(const CWalletTx& wtxIn)
  * posInBlock signals or by checking mempool presence when necessary.
  */
-bool CWallet::AddToWalletIfInvolvingMe(const CTransaction& tx, const CBlockIndex* pIndex, int posInBlock, bool fUpdate)
+bool CWallet::AddToWalletIfInvolvingMe(const CTransactionRef& ptx, const CBlockIndex* pIndex, int posInBlock, bool fUpdate)
```
Now that `AddToWalletIfInvolvingMe()` is called with `pindexBlockConnected` set to NULL when the transaction isn't in a block, can we remove the magic -1 value for `posInBlock` (and change the `if (posInBlock != -1)` into `if (pindexBlockConnected != NULL)`)?

TheBlueMatt replied:
Sure, go for it. Still gonna have to pass in two parameters either way, sadly :/.

jnewbery reviewed (commented):
concept ACK, and weak utACK (although I'm not familiar enough with the interface to be confident that I haven't missed anything).

One nit that can be addressed in a future PR.

jtimon commented:
Nice to learn about 'override'
